### PR TITLE
feat(version): add support for changelog

### PR DESCRIPTION
**Pull request by akphi**
**What's the problem this PR addresses?**
This PR aims to introduce a fairly basic changelog generation feature for version plugin
Fixes https://github.com/yarnpkg/berry/issues/1510

**How did you fix it?**
I followed the guidance listed in https://github.com/yarnpkg/berry/issues/1510#issuecomment-755584962 and referred to the approach and the UX of [changesets](https://github.com/atlassian/changesets).

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

**TODOs** - will find some time to work on these, I'm pretty slammed this week 😭 
- [ ] Support another format for a changeset (`markdown` instead of `yaml`) - right now I just add a `changelog` field to the existing `yaml` file.
- [ ] Generate and add changelog entries to `CHANGELOG.md` file.

**Things to consider** - @arcanis 
- [ ] 💬 Support multi changelog entries with respect to each workspace (more discussion needed - see https://github.com/yarnpkg/berry/issues/1510)
- [ ] Do we keep changelog when there is no accepted changelog (edge case: first there was some accepted change, they added some changelog and then they changed their mind and make the changeset unaccepted)?
- [ ] Do we need to show any placeholder text when the changelog input is empty?

### CODE REVIEW

**Comment by akphi** [comment]
@arcanis The more I work on this the more I realize we should modify (or have a flag for) version check -i so that we can scan the differences between upstream/master and the current branch and bring up interactive mode. Right now, we only based on changes in `git` working directory; this means the changeset is only for current the commit. But a lot of time I just code, commit, and want to make one commit for changelogs. Also, this is conflicting with the behavior of yarn version check -i where it allows to edit available changeset if available. I audaciously went ahead and tried to use --octopus for merge-base check so that now as long as your PR differs from upstream/master version check -i will show something. See the doc for merge-base --octopus and commit 7c36d51. We need to modify this 
https://github.com/yarnpkg/berry/blob/master/packages/plugin-version/sources/versionUtils.ts#L40-L42

Like I said, I could imagine if this is intentional, it means the changeset is per commit. But would be nice to have changeset per PR. Let me know if I completely misunderstood things here 😄 But eager to hear what you think!

**Review by arcanis — COMMENTED** [review]
Looks nice! Do you know if Ink supports a way to display an input box with a series of dots, instead of taking three rows? I quickly checked but didn't find anything 🤔 cc @vadimdemedes

File: .yarn/versions/3d70a9bb.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+releases:
+  "@yarnpkg/plugin-version": minor
```
[the remark this answers is not in the document]
Comment (akphi) [reply .yarn/versions/3d70a9bb.yml:2]: @arcanis just in the draft mode now, I will add it when I add support for changelog generation next. Will gladly be the first try-er of this feature 😄

**Comment by akphi** [comment]
> Looks nice! Do you know if Ink supports a way to display an input box with a series of dots, instead of taking three rows? I quickly checked but didn't find anything 🤔 cc @vadimdemedes

I don't know what you meant by `a series of dots`. I thought to give it a clear visual distinction, I would use some sort of bordered box, and so I looked at their [guide on border styles](https://www.npmjs.com/package/ink#borders)

**Comment by arcanis** [comment]
I meant something a bit like this: https://manaflair.github.io/mylittledom/demo/#complex-form

This way it would take a bit less vertical space. Not critical, but if there's a way ... 😄

**Comment by arcanis** [comment]
> Right now, we only based on changes in `git` working directory; this means the changeset is only for current the commit. But a lot of time I just code, commit, and want to make one commit for changelogs.

Just saw this - I'm not sure I follow. We fetch the list of changed files between master and the current HEAD, but that list includes any commit between the two of them (otherwise I wouldn't get the right result when running the edit, as you mentioned, but it works fine). The workflow you mentioned should already work 🤔

**Comment by akphi** [comment]
> I meant something a bit like this: https://manaflair.github.io/mylittledom/demo/#complex-form

This way it would take a bit less vertical space. Not critical, but if there's a way ... 😄

@arcanis awesome work there! Of course, I wouldn't mind aiming for the best possible way. I myself don't link the current border-box because when you resize it looks bad (until you type the next character - or have some `useEffect` on resize - we'll see about that).

I have looked at the doc for `ink` and it doesn't seem to me that there's a good way to achieve what you asked for (a very small comment is `...........` is a bit unpleasant to the eyes and if the user end their sentences in and actual period, that's confusing, so we need to give those fillers `..........` a more subtle color, such as grey).

Doing that way, we also have to think about text-wrapping. Looks like for `@manaflair/mylittledom`, text-wrapping is disabled and we only show truncated text. `ink` `<Text>` support this, but `ink-text-input` `<TextInput>` does not.

Also, while we're here, there's a problem with multi-line changelog. Since we block on `enter` key right now to save action, we can't really use the `return` key, on top of that, there's a `Node` [issue](https://github.com/nodejs/node/issues/2890) that prevents us from using `alt/option + enter` to create a new line 😭 - we might want to revisit our decision to use `enter` as hotkey to trigger save

Anyway, I'm all over the place here. We'll see what I can do.

**Comment by akphi** [comment]
> Just saw this - I'm not sure I follow. We fetch the list of changed files between master and the current HEAD, but that list includes any commit between the two of them (otherwise I wouldn't get the right result when running the edit, as you mentioned, but it works fine). The workflow you mentioned should already work 🤔

Just to be on the same page (because I'm so new to contributing as well), my workflow is:
1. Create a fork of `akphi/berry` of `yarn/berry`
2. Clone `akphi/berry` onto my Mac. Set `yarn/berry` as `upstream` remote
3. Work on `akphi/berry` `master` branch locally. Commit directly on `master` here
4. Push changes to my `akphi/berry` `master` branch
5. Create PR and pull changes from `akphi/berry` `master` to `yarn/berry` `master`.

So in this scenario:
`HEAD` = local `akphi/berry/master`
`master` = local `akphi/berry/master`
`origin/master` = origin `akphi/berry/master`
`yarn/berry/master` = `upstream/master`

According to this when we call `git merge-base HEAD master origin/master upstream/master` (which is our default `changesetBaseRefs`) https://github.com/yarnpkg/berry/blob/84627d8dbfa7cc9f9f8e5c22ecf0683aa8e7aeb6/packages/plugin-version/sources/versionUtils.ts#L37-L40

We will get HEAD as the base, instead of the commit where I fork `yarn/berry`. See [discussion](https://git-scm.com/docs/git-merge-base#_discussion); whereas what will result in the merge base commit I desire is to call the above command with `--octopus` [flag](https://git-scm.com/docs/git-merge-base#Documentation/git-merge-base.txt---octopus)

**Comment by arcanis** [comment]
> * Work on `akphi/berry` `master` branch locally. Commit directly on `master` here
> * Push changes to my `akphi/berry` `master` branch
> * Create PR and pull changes from `akphi/berry` `master` to `yarn/berry` `master`.

I think that's the problem. The idea (which perhaps should be better documented!) is that when working on a feature you first create a feature branch. Typically in this repo we call them `username/subject`, for instance `mael/esbuild`. Then when you run `yarn version check`, we check what are the changes in the current feature branch *compared to master*. As a result, if you run this command on master, you'll only get the changes that haven't been committed, hence the behavior you've noticed.

To sum up, a better experience would be to:

- `git clone` your repo
- `git checkout -b my/feature-name` to create a new feature branch
- Make a few commits, as many as you want
- `yarn version check -i` which will collect the version upgrades
- Want to make a few other commits? No worry, keep doing it as you want.
- You can also keep running `yarn version check -i`, it'll keep using the same file for the rest of the feature branch lifetime.
- When you're ready, make a PR and merge into master.

**Comment by akphi** [comment]
@arcanis I see, so to accommodate for the workflow that I mentioned (i.e. develop on fork `master` branch directly), I suppose we simply change `changesetBaseRefs: upstream/master`. I will revert that commit. Thanks!

**Comment by akphi** [comment]
Sorry to anyone who's waiting on this for the delay, I will come back to this in a week or two due to a tight schedule.

**Comment by fmal** [comment]
@akphi looks wonderful - would love to use this feature one day in my projects. Any chance you find time to finalize this?

**Comment by akphi** [comment]
@fmal I totally forgot about this as my original motivation was to make this work better with [changesets](https://github.com/atlassian/changesets) and I managed to tinker a bit to make it work better there. I will find time to make this happen in hopefully a week or 2. Sorry for keeping everyone waiting!

**Comment by emilio-martinez** [comment]
Hi @akphi 👋  thanks for tackling this — much needed feature. So is your latest approach to just leverage changesets under the hood?

**Comment by akphi** [comment]
> Hi @akphi 👋 thanks for tackling this — much needed feature. So is your latest approach to just leverage changesets under the hood?

@emilio-martinez The only real blocker for me to start all of this is that I cannot use `changesets` with `yarn@2` as `changesets` does [not understand the `workspace` protocol so during publish phase](https://github.com/atlassian/changesets/issues/432), it ends up publishing a bad `package.json` file to NPM. Luckily, there's a workaround [there](https://github.com/atlassian/changesets/pull/428). So this feature is no longer as much needed by me anymore.

Regardless, I'll find some time to come back to this. I think since the last time I was discussing with @arcanis I got a pretty good understanding of what it takes.

**Comment by akphi** [comment]
I'm going to close this PR here. We have moved on to use `changesets` for our project so I have lost my initial motivation to work on this feature. Really sorry about that. I will close this to make clear my intention of leaving it for someone else to pursue.

I'm still following the original thread https://github.com/yarnpkg/berry/issues/1510 and still believe being able to align `yarn` changelog with tools like `changesets` is a sensible move. As such, if someone else wants to pick this back up and build on top of my existing code, I'm happy to discuss what I have in mind.

**Comment by jj811208** [comment]
Thanks to @akphi's efforts, I think the requirement of this feature seems to be more clear after this PR.

I've been working on this issue for a few days now, but since I didn't join the discussion, my understanding of all this comes from #1510, there's no consensus on this feature yet, right?

I have summarized some insights and would like to ask you if I am on the right track🥸.

---
1.
1 PR = 1 version file mapping do not change (at least in the first iteration)

2.
> Implement an input box in yarn version check -i so that users can explain what their change is about. For the prototype an single input box is enough. Later on we may want one input per workspace, but not in the first iteration.

After working on it, I think one input box is not enough, because version files may only be partially iterated over in Yarn.

For example:

```
---
"@yarnpkg/cli": minor
"@yarnpkg/core": minor
---

Does something new
```

`cd packages/yarnpkg-core && yarn version apply`

```diff
---
"@yarnpkg/cli": minor
"@yarnpkg/core": minor
---

Does something new
```

We have no way of knowing which package this changelog is from.

But here comes the problem, I think there are multiple input boxes in the "yarn version check -i", which seems to be bad for DX, also, Ink doesn't seem to have a good textarea component.

So I want "yarn version check-i" to not pay attention to updating the changelog.

3.
> Change the generated version files to be closer from what Changeset generates. In particular, switch to markdown (but no need for a full markdown parser, I think it would be overkill). This will allow us to share more tools between our projects, which would let us use their GitHub bot, etc.

I think the conversion to markdown is a very good practice, because markdown is quite suitable for manual editing changelog.

However, if we consider this matter together with the first and second points, how to use the ecosystem of changeset will be a difficult problem.

4.
> change logs production and formatting do not belong to a package manager,

I think it makes sense that maybe yarn shouldn't care about this and just go manage everything in `package.json` (dependencies, licenses, versions...etc.)

But as a loyal yarn user, I think it would be great if yarn version deferred could support changelog🤪.

So I can convince myself that yarn is a package manager, and package manager is related to package version, and package version is related to changelog, so yarn and changelog are related.

Maybe it should be maintained by a third party? I'm not so sure, But given the context, I would expect it to be a plugin independent of yarn version plugin,

partly to make it easier for it to be extracted from the yarn repo (if needed in the future),

partly to avoid huge conflicts (the yarn team seems to be rewriting the yarn version plugin https://github.com/yarnpkg/berry/pull/4867)

In addition, it seems to be easier to integrate changeset's ecosystem when it is independent.

The tentative idea is 

adding some hooks to yarn version plugin.

1 PR = 1 version file mapping.
1 version file = 0...* changelog mapping.

changelog is manually edited by the user.

only works when version deferred

---

I'm really not very good at English and I'm still learning, so I'm sorry if I caused you any difficulty in reading

**Comment by akphi** [comment]
@jj811208 Whew, that's quite a lot. I'm afraid I only have answers for a few points though.

Regarding points 1 and 2, I think when user wants to create a new changeset, showing only 1 input box makes sense. Overall, I often find myself in 2 situations:
1. For a change, I need to add multiple changesets, each changeset corresponds to 1 package and has different messages
2. For a change, I want to use the same changeset for multiple packages and with the same message

With that, I feel like showing one input box is the only way to address these needs. If one wish to create multiple changesets like in case (1) feel free to call the command multiple times

Regarding point 4, I'm not sure about what you said, I think one of the requirements is when call `yarn version`, we will create a commit (or staging changes) which bumps the version in packages' `package.json` and also, remove the `changesets` and aggregate them into `CHANGELOG.md` file(s). As such, I'm not sure if it makes sense to keep those separate.

Anyhow, I think you have some nice ideas there, and some I can't really answer 😅 , so for the benefit of everyone, let's bring this discussion back to https://github.com/yarnpkg/berry/issues/1510 if possible. Thanks!

**Comment by jj811208** [comment]
Thank you for your reply!

I feel a lot of uncertainty with this initial idea, and I'm worried that it will be communicated to too many people.

But when I'm ready I will!
